src/ducks/subject.js

```diff
@@ -54,57 +67,118 @@ const subjectReducer = (state = initialState, action) => {
   }
 };
 
-const prepareForNewSubject = (dispatch, subject) => {
-  dispatch(resetAnnotations());
-  dispatch(createClassification(subject));
-};
-
-const fetchSubject = (initialFetch = false) => {
+/*  Fetches a Zooniverse subject from Panoptes.
+    - subjectId: OPTIONAL. ID of the subject, as a string. e.g.: "1234"
+        If unspecified (undefined), fetches from list of queued subjects.
+   */
+const fetchSubject = (subjectId = null) => {
   return (dispatch, getState) => {
-    if (initialFetch && getState().subject.status !== SUBJECT_STATUS.IDLE) return;
     const workflow_id = getState().workflow.id;
+    
+    if (!workflow_id) {
+      console.error('ducks/subjects.js fetchSubject() error: no Workflow ID');
+      return;
+    }
 
-    dispatch({ type: FETCH_SUBJECT });
-
-    const subjectQuery = { workflow_id };
+    //Fetch Specific Subject
+    if (subjectId) {
+      
+      //Store update: enter "fetching" state.
+      dispatch({ type: FETCH_SUBJECT, id: subjectId });
 
-    const fetchQueue = () => {
-      apiClient.type('subjects/queued').get(subjectQuery)
-        .then((queue) => {
-          const currentSubject = queue.shift();
+      //Asynchronous action
+      return apiClient.type('subjects').get(subjectId)
+        .then((currentSubject) => {
+          //Store update: enter "success" state and save fetched data.
           dispatch({
             currentSubject,
             id: currentSubject.id,
-            queue,
+            queue: [],
             type: FETCH_SUBJECT_SUCCESS,
             favorite: currentSubject.favorite || false
           });
 
-          prepareForNewSubject(dispatch, currentSubject);
+          //onSuccess(), prepare for a new subject.
+          dispatch(prepareForNewSubject(currentSubject));
         })
-        .catch((err) => {
-          console.error('ducks/subject.js fetchSubject() error: ', err);
+
+        .catch((err)=>{
+          //Store update: enter "error" state.
           dispatch({ type: FETCH_SUBJECT_ERROR });
         });
-    };
-
-    if (!getState().subject.queue.length) {
-      fetchQueue();
+    
+    //Fetch Next Subject In Queue
     } else {
-      const currentSubject = getState().subject.queue.shift();
-      dispatch({
-        currentSubject,
-        id: currentSubject.id,
-        queue: getState().subject.queue,
-        type: FETCH_SUBJECT_SUCCESS,
-        favorite: currentSubject.favorite || false
-      });
+      
+      const subjectQuery = { workflow_id };
+
+      // TODO What if the fetched queue is empty?
```

Comment:
Are you referring to error handling here? I think this would only occur if there weren't any subjects associated with in workflow.

Reply:
Yup! This wouldn't happen in Scribes of the Cairo Geniza (unless something really effs up) but this comment came part of the "generalised CFE library" code, and I wanted to make sure that some day we'll be able to account for the situation when the app tries to get a queue from 1. a project with no subject sets, 2. a project with subjects sets that have no subjects, 3. a project with pretty much fully classified subjects OK there's nothing to do what do you do now.  The _generalised_ fetchSubject() feature shouldn't just return an empty array, it should also raise a flag of some sort to indicate, "Hey, the fetch was successful... but we just have nothing, yo."